### PR TITLE
fix(schema-engine): Ignore views for MongoDB Schemas

### PR DESCRIPTION
**Pull request by christianledgard**
Related to [#18424](https://github.com/prisma/prisma/issues/18424) and [#17006](https://github.com/prisma/prisma/issues/17006) issues.

**Proposed trivial approach**: While iterating the collections, if we detect a View, we continue to the next collection.

### CODE REVIEW

**Review by Jolg42 — CHANGES_REQUESTED** [review]
Blocking this because I tried it and while the error message is gone it brings unexpected changes

I ran `PRISMA_SCHEMA_ENGINE_BINARY=/Users/j42/Dev/prisma-engines/target/debug/schema-engine npx prisma db pull --force` which ran successfully.

Though check the output
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = "mongodb://root:prisma@localhost:27018/hey?authSource=admin"
}

type System.viewsPipeline {
lookup System.viewsPipeline$lookup? @map("$lookup")
project System.viewsPipeline$project? @map("$project")
unwind String? @map("$unwind")
}

type System.viewsPipeline$lookup {
as String
foreignField String
from String
localField String
}

type System.viewsPipeline$project {
id Int @map("_id")
bio String
email Int
name Int
}

model Plan {
  id String @id @default(auto()) @map("_id") @db.ObjectId
}

model ho {
  id String @id @default(auto()) @map("_id") @db.ObjectId
}

model system_views {
  id     String @id @map("_id")
  pipeline System.viewsPipeline[]
  viewOn String

  @@map("system.views")
}
```
The `type System.*` are unexpected (and also not valid syntax, no period is allowed in the name).
So it looks like some logic needs to be adapted to ignore introspecting the `system` collection itself?

**Comment by Jolg42** [comment]
I pushed a fix. I'll ask a colleague about how we want to test this.

**Comment by Jolg42** [comment]
Note: I verified that the test was not working before and showed the same issues before the fix.

**Comment by janpio** [comment]
> I'm not sure why Clippy errors on unrelated files?

That's that clippy usually does 😄

**Comment by aqrln** [comment]
> I'm not sure why Clippy errors on unrelated files?

Rust 1.72 was released yesterday so probably was now updated on CI.

File: schema-engine/mongodb-schema-describer/src/lib.rs [hunk 1]
```diff
@@ -27,9 +27,21 @@ pub async fn describe(client: &mongodb::Client, db_name: &str) -> mongodb::error
...
 
+        // We need to skip views, we do not support introspecting them yet.
+        if collection_type == mongodb::results::CollectionType::View {
+            continue;
+        }
+
+        // We need to skip system collections, they are only used by MongoDB internally.
+        // https://www.mongodb.com/docs/manual/reference/system-collections/
+        if collection_type == mongodb::results::CollectionType::Collection && collection_name.starts_with("system.") {
+            continue;
```
Comment (tomhoule) [inline schema-engine/mongodb-schema-describer/src/lib.rs:42]: Do these show up because we have a view? That's a bit mysterious.

**Review by tomhoule — APPROVED** [review]
The fixes and test look good :100:

File: schema-engine/mongodb-schema-describer/src/lib.rs (hunk 1 above)
Comment (Jolg42) [reply schema-engine/mongodb-schema-describer/src/lib.rs:42]: Yes, when there is a view inside the database MongoDB stores some information inside that same database as a `system.views` (invisible in GUI) collection

**Review by Jolg42 — APPROVED** [review]

**Comment by Jolg42** [comment]
Note: this is now available in `5.3.0-dev.9` if you want to try it out.
It will be released in the official 5.3.0 version on Tuesday, September, 12th.

**Pull request merged**